opentelemetry/src/attributes/set.rs

```diff
@@ -0,0 +1,366 @@
+        Value::String(_) => 4,
+        Value::Array(a) => match a {
+            Array::Bool(_) => 5,
+            Array::I64(_) => 6,
+            Array::F64(_) => 7,
+            Array::String(_) => 8,
+        },
+    }
+}
+
+impl PartialEq for HashKeyValue {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.key == other.0.key
+            && match (&self.0.value, &other.0.value) {
+                (Value::F64(f), Value::F64(of)) => OrderedFloat(*f).eq(&OrderedFloat(*of)),
+                (Value::Array(Array::F64(f)), Value::Array(Array::F64(of))) => {
+                    f.len() == of.len()
+                        && f.iter()
+                            .zip(of.iter())
+                            .all(|(f, of)| OrderedFloat(*f).eq(&OrderedFloat(*of)))
+                }
+                (non_float, other_non_float) => non_float.eq(other_non_float),
+            }
+    }
+}
+
+impl Eq for HashKeyValue {}
+
+static EMPTY_SET: Lazy<Arc<InternalAttributeSet>> =
+    Lazy::new(|| Arc::new(InternalAttributeSet::new(Vec::with_capacity(0))));
+
+#[derive(Eq, PartialEq, Debug)]
+struct InternalAttributeSet {
+    key_values: Vec<HashKeyValue>,
+    hash: u64,
+}
+
+impl InternalAttributeSet {
+    fn new(mut values: Vec<HashKeyValue>) -> Self {
+        values.sort_unstable();
+        let mut hasher = DefaultHasher::new();
+        values.iter().fold(&mut hasher, |mut hasher, item| {
+            item.hash(&mut hasher);
+            hasher
+        });
+
+        InternalAttributeSet {
+            key_values: values,
+            hash: hasher.finish(),
+        }
+    }
+}
+
+impl<I: IntoIterator<Item = KeyValue>> From<I> for InternalAttributeSet
+where
+    <I as IntoIterator>::IntoIter: DoubleEndedIterator + ExactSizeIterator,
+{
+    fn from(value: I) -> Self {
+        let iter = value.into_iter();
+        let mut seen_keys = HashSet::with_capacity(iter.len());
+        let vec = iter
+            .into_iter()
+            .rev()
+            .filter_map(|kv| {
+                if seen_keys.contains(&kv.key) {
+                    None
+                } else {
+                    seen_keys.insert(kv.key.clone());
+                    Some(HashKeyValue(kv))
+                }
+            })
+            .collect::<Vec<_>>();
+
+        InternalAttributeSet::new(vec)
+    }
+}
+
+impl Hash for InternalAttributeSet {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_u64(self.hash)
+    }
+}
+
+/// A unique set of attributes that can be used as instrument identifiers.
+///
+/// Cloning of an attribute set is cheap, as all clones share a reference to the underlying
+/// attribute data.
+///
+/// This must implement [Hash], [PartialEq], and [Eq] so it may be used as
+/// HashMap keys and other de-duplication methods.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct AttributeSet(Arc<InternalAttributeSet>);
+
+impl<'a, I: IntoIterator<Item = &'a KeyValue>> From<I> for AttributeSet
```

Comment:
It's a little unfortunate to require this to always be references from a caller perspective. One alternative could be to make a new trait so this can accept both iterators over references and owned values? e.g. 

```rust
trait ToKv {
    fn to_kv(self) -> KeyValue;
}

impl ToKv for KeyValue {
    fn to_kv(self) -> KeyValue {
        self
    }
}

impl ToKv for &'_ KeyValue {
    fn to_kv(self) -> KeyValue {
        self.clone()
    }
}

impl<'a, KV, I> From<I> for AttributeSet
where
    KV: ToKv,
    I: IntoIterator<Item = KV>,
    <I as IntoIterator>::IntoIter: DoubleEndedIterator + ExactSizeIterator,
{
    fn from(values: I) -> Self {
        AttributeSet(Arc::new(InternalAttributeSet::from(
            values.into_iter().map(ToKv::to_kv),
        )))
    }
}
```

That would allow for a greater variety of arguments 

```rust
let _ = AttributeSet::from(vec![KeyValue::new("owned", true)]);
let _ = AttributeSet::from(&[KeyValue::new("owned", false)]);
```